Add Apache Phoenix connector

vincentpoon opened the pull request:
https://phoenix.apache.org/

Ref https://github.com/prestodb/presto/pull/12111

martint commented:
Thanks @vincentpoon. We're going to start looking at this shortly. In the meantime, please make sure to sign the CLA so we can get your code merged.

vincentpoon commented:
FYI @lhofhansl @hustnn @aertoria @JamesRTaylor

martint commented:
@cla-bot check

kokosing commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixPageSource.java:
```diff
@@ -0,0 +1,407 @@
+        nanoEnd = System.nanoTime();
+    }
+
+    private RuntimeException handleSqlException(Exception e)
```
Exception is not a SqlException. At least not always.

kokosing commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixPageSourceProvider.java:
```diff
@@ -0,0 +1,54 @@
+    {
+        List<PhoenixColumnHandle> phoenixCols = columns.stream().map(column -> (PhoenixColumnHandle) column).collect(Collectors.toList());
+        if (PhoenixMetadata.getPkHandle(phoenixCols).isPresent()) {
+            return new PhoenixUpdatablePageSource(phoenixClient,
```
each argument in separate line or all in one line.

kokosing commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,258 @@
+        final QueryPlan queryPlan = getQueryPlan(context, configuration);
+        final List<KeyRange> allSplits = queryPlan.getSplits();
+        final List<InputSplit> splits = generateSplits(queryPlan, allSplits, configuration);
+        return splits;
```
inline

kokosing commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/TestingPhoenixServer.java:
```diff
@@ -0,0 +1,99 @@
+        }));
+    }
+
+    public void close()
```
`@Override`

kokosing commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/TestPhoenixDistributedQueries.java:
```diff
@@ -0,0 +1,134 @@
+
+        assertQuery("SELECT * FROM test_insert", query);
+
+        assertUpdate("INSERT INTO test_insert (orderkey) VALUES (-1)", 1);
```
Can you run this insert twice?

vincentpoon replied:
Running twice is an Upsert in Phoenix.  I added the extra insert, doesn't impact the test.

findepi replied:
will the second INSERT also return `1` as the number of rows changed?

vincentpoon replied:
Yes, the second INSERT updates 1 row (since it's a Phoenix Upsert).

kokosing commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/PhoenixQueryRunner.java:
```diff
@@ -0,0 +1,154 @@
+                break;
+        }
+
+        LOG.info("Running import for %s", target, sql);
```
debug

kokosing commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/PhoenixQueryRunner.java:
```diff
@@ -0,0 +1,154 @@
+        LOG.info("%s", sql);
+        long start = System.nanoTime();
+        long rows = queryRunner.execute(session, sql).getUpdateCount().getAsLong();
+        LOG.info("Imported %s rows for %s in %s", rows, target, nanosSince(start));
```
We should leverage Presto execution statistics for this. I see no point in logging this.

vincentpoon commented:
Thanks for the initial review @kokosing , I've just pushed fixes to most of your suggestions and more cleanup.

vincentpoon commented:
@combineads Can you submit the CLA?  https://github.com/prestosql/cla
Tried emailing you, but didn't get a response.

vincentpoon commented on presto-docs/src/main/sphinx/connector/phoenix.rst:
```diff
@@ -0,0 +1,137 @@
+.. code-block:: none
+
+    connector.name=phoenix
+    connection-url=jdbc:phoenix:host1,host2,host3:2181:/hbase
```
That is a standard zookeeper quorum connection string - all the hosts should be on the same port (default 2181).
I didn't look deeply into base-jdbc, but it seems in the end most of the implementation would be similar (still need SplitManager, PageSource, etc).  There are some things unique to Phoenix that aren't standard SQL, like Upsert.
We aren't actually sending SQL commands to the HBase cluster here.  We are merely using the Phoenix client query planner to construct HBase scans, then use those scans in the splits.  So it's a weird hybrid.

hustnn replied:
If using JDBC, the calculation will be executed internally in phoenix/hbase.
The connector here is just telling presto how to retrieve the raw data split (scan in hbase), the calculation logic is done in presto cluster itself.

Is my understanding right? If it is like this, which approach will be better?

vincentpoon commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,119 @@
+        }
+    }
+
+    public String getConnectorId()
```
It is possible to have multiple phoenix connector catalogs connecting to different HBase clusters, so I think we still need this.  If someone wants to query catalogs separately, for instance, and/or do joins to federate.  That's actually one of our initial use cases.  This is the same as the other connectors.

A future enhancement would be to also support multiple Phoenix clusters within the same connector catalog.  But we would still want the above.

vincentpoon commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixPageSink.java:
```diff
@@ -0,0 +1,192 @@
+        String vars = Joiner.on(',').join(nCopies(handle.getColumnNames().size(), "?"));
+        // ON DUPLICATE KEY UPDATE counter1 = counter1 + 1, counter2 = counter2 + 1;
+        StringBuilder sql = new StringBuilder()
+                .append("UPSERT INTO ")
```
Upsert is the only Phoenix way to Insert.  Phoenix doesn't distinguish between insert and update.

vincentpoon commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixPageSource.java:
```diff
@@ -0,0 +1,407 @@
+        PhoenixConfigurationUtil.setInputClass(conf, PhoenixDBWritable.class);
+        TaskAttemptContextImpl taskContext = new TaskAttemptContextImpl(conf, new TaskAttemptID());
+        PhoenixInputSplit pInputSplit = split.getPhoenixInputSplit();
+        RecordReader<NullWritable, PhoenixDBWritable> reader = new PhoenixInputFormat<PhoenixDBWritable>().createRecordReader(pInputSplit, taskContext);
```
We are trying to reuse the Phoenix MR framework.  We use the record reader there to iterate over the split.

vincentpoon commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixPageSource.java:
```diff
@@ -0,0 +1,407 @@
+        }
+        closed = true;
+
+        // use try with resources to close everything properly
```
Some of these are only AutoCloseable, not Closeable (which Closer requires). So this seems to be the right way to close.

findepi commented:
@vincentpoon please squash your commits into single commit. Place yourself as the commit author (`git commit --amend --reset-author`) and use this as the message:
```
Add Apache Phoenix connector

Based on @combineads (Byunghwa Yun <combine@combineads.co.kr>)
published at https://github.com/prestodb/presto/pull/12111.
```

vincentpoon commented:
@findepi Done, thanks for the suggestion

hustnn commented on presto-docs/src/main/sphinx/connector/phoenix.rst:
```diff
@@ -0,0 +1,137 @@
+
+    connector.name=phoenix
+    connection-url=jdbc:phoenix:host1,host2,host3:2181:/hbase
+    connection-properties=phoenix.schema.isNamespaceMappingEnabled=true
```
There is a mapping between Hbase table name and phoenix table name

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSessionProperties.java:
```diff
@@ -0,0 +1,70 @@
+        sessionProperties = ImmutableList.of(
+                stringProperty(
+                        DUPLICATE_KEY_UPDATE_COLUMNS,
+                        "A comma-delimited list of Presto columns that the row will be updated.",
```
> that the row will be updated.

grammar?

i sincerely don't understand what it does. But i'm not pheonix user.

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSessionProperties.java:
```diff
@@ -0,0 +1,70 @@
+public final class PhoenixSessionProperties
+{
+    private static final String DUPLICATE_KEY_UPDATE_COLUMNS = "duplicate_key_update_columns";
+    private static final Splitter DUPLICATE_KEY_UPDATE_COLUMNS_SPLITTER = Splitter.on(" and ").trimResults();
```
i would expect `","` or something. Is it something phoenix's customary ?

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,252 @@
+                if (splitByStats) {
+                    for (Scan aScan : scans) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Split for  scan : " + aScan + "with scanAttribute : " +
```
Use `%s` placeholders.
Then, is `if (log.isDebugEnabled())` required?

electrum replied:
To expand on the above comment, the Airlift logger uses standard Java string formatting. An easy way to fix these in IntelliJ is to put the cursor on the `+`, hit alt-enter for quick fix, choose *"Replace with String.format()"*, then remove the outer `format()` call.

The string format is only evaluated if the log level is enabled. Thus, assuming all the calls like `aScan.getBatch()` are cheap, there's no need to conditionally call the log method. Of course, this would not be true for things like `Bytes.toStringBinary()`, since that would be evaluated every time.

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,252 @@
+                boolean splitByStats = PhoenixConfigurationUtil.getSplitByStats(config);
+
+                if (splitByStats) {
+                    for (Scan aScan : scans) {
```
```suggestion
                    for (Scan scan : scans) {
```

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,252 @@
+                HBaseFactoryProvider.getHConnectionFactory().createConnection(config)) {
+            RegionLocator regionLocator = connection.getRegionLocator(TableName.valueOf(qplan.getTableRef().getTable().getPhysicalName().toString()));
+            long regionSize = -1;
+            final List<InputSplit> psplits = Lists.newArrayListWithExpectedSize(splits.size());
```
```suggestion
            final List<InputSplit> splits = Lists.newArrayListWithExpectedSize(splits.size());
```

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,252 @@
+    private List<InputSplit> generateSplits(final QueryPlan qplan, final List<KeyRange> splits, Configuration config)
+            throws IOException
+    {
+        Preconditions.checkNotNull(qplan);
```
```suggestion
        requireNonNull(queryPlan, "queryPlan is null");
```

(variable name, use `java.util.Objects#requireNonNull(T, java.lang.String)`)

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,252 @@
+    {
+        Preconditions.checkNotNull(context);
+        try {
+            final String txnScnValue = configuration.get(PhoenixConfigurationUtil.TX_SCN_VALUE);
```
```suggestion
            final String transactionScanValue = configuration.get(PhoenixConfigurationUtil.TX_SCN_VALUE);
```
?

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,252 @@
+        }
+    }
+
+    private QueryPlan getQueryPlan(final JobContext context, final Configuration configuration)
```
per code style, we don't annotate method params as `final`

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixTableProperties.java:
```diff
@@ -0,0 +1,265 @@
+    public PhoenixTableProperties(TypeManager typeManager)
+    {
+        tableProperties = ImmutableList.of(
+                stringProperty(ROWKEYS,
```
```suggestion
                stringProperty(
                ROWKEYS,
```

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/QueryBuilder.java:
```diff
@@ -0,0 +1,364 @@
+                    statement.setTime(i + 1, new Time(unpackMillisUtc((long) typeAndValue.getValue())));
+                }
+                else if (type.equals(TimestampType.TIMESTAMP)) {
+                    statement.setTimestamp(i + 1, new Timestamp((long) typeAndValue.getValue()));
```
this code should check `io.prestosql.spi.connector.ConnectorSession#isLegacyTimestamp`

This code is not correct  regardless of TIMESTAMP semantics
(unless JVM zone is UTC or equal to presto user session zone)

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/QueryBuilder.java:
```diff
@@ -0,0 +1,364 @@
+                    statement.setDate(i + 1, new Date(UTC.getMillisKeepLocal(DateTimeZone.getDefault(), millis)));
+                }
+                else if (type.equals(TimeType.TIME)) {
+                    statement.setTime(i + 1, new Time((long) typeAndValue.getValue()));
```
this code should check `io.prestosql.spi.connector.ConnectorSession#isLegacyTimestamp`

This code is not correct  regardless of TIMESTAMP semantics
(unless JVM zone is UTC or equal to presto user session zone)

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/QueryBuilder.java:
```diff
@@ -0,0 +1,364 @@
+                    statement.setTime(i + 1, new Time((long) typeAndValue.getValue()));
+                }
+                else if (type.equals(TimeWithTimeZoneType.TIME_WITH_TIME_ZONE)) {
+                    statement.setTime(i + 1, new Time(unpackMillisUtc((long) typeAndValue.getValue())));
```
This code is not correct unless JVM zone = UTC

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/QueryBuilder.java:
```diff
@@ -0,0 +1,364 @@
+                    statement.setTimestamp(i + 1, new Timestamp((long) typeAndValue.getValue()));
+                }
+                else if (type.equals(TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE)) {
+                    statement.setTimestamp(i + 1, new Timestamp(unpackMillisUtc((long) typeAndValue.getValue())));
```
This code is not correct unless JVM zone = UTC

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/QueryBuilder.java:
```diff
@@ -0,0 +1,364 @@
+                else if (isLongDecimal(type)) {
+                    int scale = ((DecimalType) type).getScale();
+                    BigInteger unscaledValue = Decimals.decodeUnscaledValue((Slice) typeAndValue.getValue());
+                    statement.setBigDecimal(i + 1, new BigDecimal(unscaledValue, scale));
```
```suggestion
                    statement.setBigDecimal(i + 1, new BigDecimal(unscaledValue, scale, new MathContext(type.getPrecision())));
```
?

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/QueryBuilder.java:
```diff
@@ -0,0 +1,364 @@
+                else if (isShortDecimal(type)) {
+                    int scale = ((DecimalType) type).getScale();
+                    BigInteger unscaledValue = BigInteger.valueOf((long) typeAndValue.getValue());
+                    statement.setBigDecimal(i + 1, new BigDecimal(unscaledValue, scale));
```
```suggestion
                    statement.setBigDecimal(i + 1, new BigDecimal(unscaledValue, scale, new MathContext(type.getPrecision())));
```
?

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/QueryBuilder.java:
```diff
@@ -0,0 +1,364 @@
+                if (!range.getLow().isLowerUnbounded()) {
+                    switch (range.getLow().getBound()) {
+                        case ABOVE:
+                            rangeConjuncts.add(toPredicate(columnName, ">", range.getLow().getValue(), type, accumulator));
```
I would feel safer if we could reuse base-jdbc QueryBuilder instead of copying the class

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java:
```diff
@@ -0,0 +1,261 @@
+            .put(VARBINARY, "VARBINARY")
+            .put(DATE, "DATE")
+            .put(TIME, "TIME")
+            .put(TIME_WITH_TIME_ZONE, "TIME")
```
remove

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java:
```diff
@@ -0,0 +1,261 @@
+            .put(TIME, "TIME")
+            .put(TIME_WITH_TIME_ZONE, "TIME")
+            .put(TIMESTAMP, "TIMESTAMP")
+            .put(TIMESTAMP_WITH_TIME_ZONE, "TIMESTAMP")
```
remove

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java:
```diff
@@ -0,0 +1,261 @@
+        return type.getTypeSignature().getBase().equals(StandardTypes.ROW);
+    }
+
+    public static boolean isDateType(Type type)
```
```suggestion
    public static boolean isDateTimeType(Type type)
```

findepi replied:
unused, remove

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java:
```diff
@@ -0,0 +1,261 @@
+        return type.getTypeSignature().getBase().equals(StandardTypes.ARRAY);
+    }
+
+    public static boolean isMapType(Type type)
```
unused, remove

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java:
```diff
@@ -0,0 +1,261 @@
+        return type.getTypeSignature().getBase().equals(StandardTypes.MAP);
+    }
+
+    public static boolean isRowType(Type type)
```
unused, remove

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java:
```diff
@@ -0,0 +1,261 @@
+
+    public static boolean isArrayType(Type type)
+    {
+        return type.getTypeSignature().getBase().equals(StandardTypes.ARRAY);
```
why not just `type instanceof ArrayType`?

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java:
```diff
@@ -0,0 +1,261 @@
+            if (((VarcharType) type).isUnbounded()) {
+                return "VARCHAR";
+            }
+            return "VARCHAR(" + ((VarcharType) type).getBoundedLength() + ")";
```
is there a limit for varchar in pheonix?

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java:
```diff
@@ -0,0 +1,261 @@
+        }
+        if (type instanceof CharType) {
+            if (((CharType) type).getLength() == CharType.MAX_LENGTH) {
+                return "CHAR";
```
Why? (`CharType.MAX_LENGTH` is not a "placeholder value")

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java:
```diff
@@ -0,0 +1,261 @@
+        }
+
+        if (type instanceof DecimalType) {
+            return type.toString().toUpperCase();
```
Don't use toString() 

```suggestion
            return format("DECIMAL(%s, %s)", ....
```

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java:
```diff
@@ -0,0 +1,261 @@
+                return DecimalType.createDecimalType(columnSize, decimalDigits);
+            case Types.CHAR:
+            case Types.NCHAR:
+                return createCharType(min(columnSize, CharType.MAX_LENGTH));
```
if `columnSize > CharType.MAX_LENGTH`, you need to represent the data as `varchar`. 
Otherwise, what would you do with the data values that are too long to fit?

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java:
```diff
@@ -0,0 +1,261 @@
+            case Types.LONGVARCHAR:
+            case Types.LONGNVARCHAR:
+                if (columnSize == 0 || columnSize > VarcharType.MAX_LENGTH) {
+                    return createUnboundedVarcharType();
```
Why `columnSize == 0`? I think Presto supports `varchar(0)` just fine.

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java:
```diff
@@ -0,0 +1,261 @@
+        }
+        else if (TIMESTAMP.equals(type)) {
+            long millisUtc = type.getLong(block, position);
+            return new Timestamp(millisUtc);
```
this code should check `io.prestosql.spi.connector.ConnectorSession#isLegacyTimestamp`

This code is not correct  regardless of TIMESTAMP semantics
(unless JVM zone is UTC or equal to presto user session zone)

findepi commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java:
```diff
@@ -0,0 +1,261 @@
+    }
+
+    public static Object getObjectValue(Type type, Block block, int position)
+            throws SQLException
```
not thrown here

findepi commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/PhoenixQueryRunner.java:
```diff
@@ -0,0 +1,148 @@
+    private static final String TPCH_SCHEMA = "tpch";
+
+    private static boolean tpchLoaded;
+    private static TestingPhoenixServer server = new TestingPhoenixServer();
```
`PhoenixQueryRunner` cannot be instantiated concurrently.
 make `server` non-static

findepi commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/TestPhoenixDistributedQueries.java:
```diff
@@ -0,0 +1,135 @@
+        @Language("SQL")
+        String query = "SELECT orderdate, orderkey, totalprice FROM orders";
+
+        assertUpdate("CREATE TABLE test_insert WITH (ROWKEYS='orderkey')AS " + query + " WITH NO DATA", 0);
```
```suggestion
        assertUpdate("CREATE TABLE test_insert WITH (ROWKEYS='orderkey') AS " + query + " WITH NO DATA", 0);
```

findepi commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/PhoenixQueryRunner.java:
```diff
@@ -0,0 +1,148 @@
+        if (!tpchLoaded) {
+            createSchema(server, TPCH_SCHEMA);
+            copyTpchTables(queryRunner, "tpch", TINY_SCHEMA_NAME, createSession(), tables);
+            tpchLoaded = true;
```
The set of tables created depends on the value of `tables` collection which first passes `!tpchLoaded` test.
Thus, the tests depend on their order of execution

findepi commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/PhoenixQueryRunner.java:
```diff
@@ -0,0 +1,148 @@
+                sql = format("CREATE TABLE %s AS SELECT * FROM %s", target, source);
+                break;
+            default:
+                sql = format("CREATE TABLE %s AS SELECT * FROM %s", target, source);
```
You could use the `switch` to define `String tableProperties` (either `""` or `"WITH (ROWKEYS ...)"`)
This way, the code would be shorter and your intent clearer

findepi commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/TestPhoenixDistributedQueries.java:
```diff
@@ -0,0 +1,135 @@
+
+        assertUpdate("INSERT INTO test_insert (pk, a) VALUES (1, ARRAY[null])", 1);
+        assertUpdate("INSERT INTO test_insert (pk, a) VALUES (2, ARRAY[1234])", 1);
+        // An array of numeric primitive types returns 0 when the value is null.
```
it should fail instead

lhofhansl replied:
Hi @findepi .
There is no "insert" or "update" in Phoenix. Everything is an upsert. So this would write a new version of a "row" with the same key.

findepi commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/TestingPhoenixServer.java:
```diff
@@ -0,0 +1,108 @@
+            throw new IllegalStateException("Can't start phoenix server.", e);
+        }
+
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
```
Don't depend on shutdown hooks for anything. Remove

findepi commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/TestingPhoenixServer.java:
```diff
@@ -0,0 +1,108 @@
+        }
+    }
+
+    private static int randomPort()
```
This will make the tests flaky. see https://github.com/prestosql/presto/pull/53

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixConnector.java:
```diff
@@ -0,0 +1,152 @@
+        try {
+            lifeCycleManager.stop();
+        }
+        catch (Exception e) {
```
It is needed, because `shutdown()` can't throw checked exceptions and `LifeCycleManager.stop()` throws `Exception`.

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixErrorCode.java:
```diff
@@ -0,0 +1,40 @@
+
+    PhoenixErrorCode(int code, ErrorType type)
+    {
+        errorCode = new ErrorCode(code + 0x0105_0000, name(), type);
```
`0x0105` is for the Thrift connector. I registered `0x0106` for the Phoenix connector here: https://github.com/prestosql/presto/wiki/Error-Codes

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,252 @@
+            }
+        }
+        catch (Exception e) {
+            log.error(String.format("Failed to get the query plan with error [%s]", e.getMessage()));
```
log-and-throw is an anti-pattern that we avoid in Presto. It results in the same error getting logged in multiple places, which is confusing. Either log or throw, but not both. In this case, it's probably best to re-throw like this:
```java
throw new PrestoException(PHOENIX_ERROR, "Failed to get the Phoenix query plan", e);

electrum replied:
Also, create separate error codes when possible. For example, `PHOENIX_QUERY_PLAN_ERROR` might be appropriate here. It helps administrators to understand why understand why queries failed, count query failures by error type, etc.

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,252 @@
+    }
+
+    // mostly copied from PhoenixInputFormat, but without the region size calculations
+    private List<InputSplit> generateSplits(final QueryPlan qplan, final List<KeyRange> splits, Configuration config)
```
Nit: don't put `final` on parameters or local variables

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplit.java:
```diff
@@ -0,0 +1,184 @@
+    {
+        ToStringHelper helper = toStringHelper(this)
+                .add("connectorId", connectorId);
+        if (catalogName != null) {
```
Rather than checking for nulls, add them unconditionally and call `omitNullValues()` at the end.

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,252 @@
+                    layoutHandle.getTupleDomain(),
+                    phoenixMetadata.getColumns(handle, false));
+
+            return new FixedSplitSource(getHadoopInputSplits(inputQuery).stream().map(split -> (PhoenixInputSplit) split).map(split -> {
```
You can do
```java
map(PhoenixInputSplit.class::cast)

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,252 @@
+                    phoenixMetadata.getColumns(handle, false));
+
+            return new FixedSplitSource(getHadoopInputSplits(inputQuery).stream().map(split -> (PhoenixInputSplit) split).map(split -> {
+                List<HostAddress> addresses = getSplitAddresses(split);
```
This variable can be inlined since it's only used once. And after that, the lambda body can become a lambda expression (no need for a block with `return`)

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,252 @@
+            }
+            addresses = ImmutableList.of(address);
+        }
+        catch (Exception e) {
```
This should be `RuntimeException` since there are no checked exceptions thrown.

The log message should probably include some information about the split, otherwise it doesn't seem useful. Also consider how often this will be logged. If it's called for each split, the log will be flooded with this message. Is there any action the administrator can take based on this log message?

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,252 @@
+
+    private final PhoenixClient phoenixClient;
+    private PhoenixMetadata phoenixMetadata;
+    private final Map<String, HostAddress> hostCache = new HashMap<>();
```
This needs to be a `ConcurrentHashMap` since the split manager is accessed by multiple threads concurrently for different queries.

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,252 @@
+    private static final Logger log = Logger.get(PhoenixSplitManager.class);
+
+    private final PhoenixClient phoenixClient;
+    private PhoenixMetadata phoenixMetadata;
```
This can be final

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,252 @@
+        List<HostAddress> addresses;
+        try {
+            String hostName = split.getLocations()[0];
+            HostAddress address = hostCache.get(hostName);
```
This can be
```java
HostAddress address = hostCache.computeIfAbsent(hostName, HostAddress::fromString);

electrum replied:
I'm not convinced we need a cache at all here. `HostAddress.fromString()` is doing basic parsing to split a string into a host and port. There is no address resolution or anything else expensive.

vincentpoon commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixConfig.java:
```diff
@@ -0,0 +1,65 @@
+public class PhoenixConfig
+{
+    private Properties connectionProperties = new Properties();
+    private String connectionUrl;
```
There's no user/pass , Phoenix only supports Kerberos

kokosing replied:
it would be nice to make it possible to configure kerberos without including (loading) hhbase-site.xml or core-site.xml files

vincentpoon replied:
You can already configure kerberos through the connection string:
jdbc:phoenix:[comma-separated ZooKeeper Quorum [:port [:hbase root znode [:kerberos_principal [:path to kerberos keytab] ] ] ]

vincentpoon commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixConfig.java:
```diff
@@ -0,0 +1,65 @@
+        return connectionProperties;
+    }
+
+    @Config("connection-properties")
```
I think having the user serialize by hand is fine, similar to what you see in JDBC urls (e.g. `jdbc:mysql://<host>:<port>/<database_name>?property1=value1&property2=value2`).  Unfortunately Phoenix doesn't take everything as a url property, so we need this separate property.
There are a ton of potential Phoenix properties, it won't scale to enumerate all of them in the class.

vincentpoon commented on presto-docs/src/main/sphinx/connector/phoenix.rst:
```diff
@@ -0,0 +1,155 @@
+-----------------------------
+
+* Only one dimensional arrays are currently supported.
+* Tables in the Phoenix default schema without namespace mapping cannot be queried
```
I'm adding support for "default" schema, so getting rid of this documentation.  Also the limitation on one dimensional arrays is a Phoenix limitation, not a limitation of this connector, so removing that as well.

vincentpoon commented on presto-phoenix/pom.xml:
```diff
@@ -0,0 +1,182 @@
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.phoenix</groupId>
+            <artifactId>presto-phoenix-client-shaded</artifactId>
```
We can maintain that shaded jar in Phoenix.

vincentpoon commented:
Still working on this :)
As you know, I based this work off someone else's old pull request.  From the review comments, I've come to realize that much of that code was copy+pasted from base-jdbc and other places.  I'm trying to rewrite large portions to fix all of that.
Thanks for the patience.  BTW, I appreciate all the comments - the codebase is a pleasure to work with due to diligence from maintainers like yourselves in keeping the code clean and sharp!

combineads commented:
@vincentpoon 
I am so sorry. I could not confirm my email due to personal circumstances. I have submitted a CLA.
Please feel free to ask me if I can help. Thank you for your effort.

vincentpoon commented:
@findepi @electrum @kokosing 

I've pushed a new revision.  Would appreciate you taking a looking.
It's a major overhaul which removed a lot of duplicate code and reuses code from base-jdbc.  To give you an idea, the patch file line count dropped from ~5500 to ~3000.  I also fixed up the patch with all the prior review suggestions.

At the core, what we're doing is pretty simple - generate Phoenix splits (HBase scans) and wrap them in a ResultSet, and from then on treat each split as if it were from a jdbc connector.  I did have to drop a few things, though:  1) ARRAY type, which I hope to complete as a broader patch in #317, 2) DELETE, which doesn't appear to be implemented for JDBC yet, and 3) Phoenix ON DUPLICATE KEY support.

Other notables:  Connection properties are now read from a separate hbase-site.xml, which is the way it's typically done in Phoenix deployments.  The phoenix.rst is cleaned up - I shortened all the descriptions and provided a link instead, as all the properties would generally be familiar to Phoenix/HBase users and more information can be found on the Phoenix/HBase websites anyways.

Thanks!
FYI @lhofhansl

kokosing commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,283 @@
+            };
+        }
+        catch (IOException | InterruptedException e) {
+            throw new PrestoException(PHOENIX_QUERY_ERROR, "Error while setting up Phoenix ResultSet", e);
```
Interrupted exception is not an error. Other than that why `InterruptedException` is thrown here.

kokosing commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,283 @@
+                if (typeHandle.getColumnSize() == 0) {
+                    return Optional.of(varcharColumnMapping(createUnboundedVarcharType()));
+                }
+                else {
```
redundant else block

kokosing commented on presto-phoenix/pom.xml:
```diff
@@ -0,0 +1,174 @@
+
+        <dependency>
+            <groupId>io.prestosql</groupId>
+            <artifactId>presto-base-jdbc</artifactId>
```
I am not sure this relation of this connector with `presto-base-jdbc`. Can you instead copy code you need and adjust it for phoenix needs instead of using `presto-base-jdbc` partially. Then maybe you wouldn't need to map `PhoenixRecordReader` to `ResultsSet` but you could use `PhoenixRecordReader` directly to produce data.

I am afraid that such usage of `presto-base-jdbc` would make `presto-base-jdbc` to be harder to maintain and evolve.

What do you think?

vincentpoon replied:
In your first review, you wrote 
>Also, why not to use base-jdbc connector to create phoenix-connector? We could extend base-> jdbc connector if needed.

:).  The fact is, Phoenix is a SQL layer on top of a NoSQL store, so it is a weird hybrid, so I can see arguments for either way.
Phoenix implements most of the jdbc interface, and supports reading through a jdbc `ResultSet`.  The code in base-jdbc does all the Type translation for us here - e.g. `StandardColumnMapping` and the `Read/WriteFunction` can all be reused here. 

Copying the code was the approach the first PR tried.  It was a lot uglier IMHO and wouldn't inherit improvements to base-jdbc.

Btw I removed usage of `PhoenixRecordReader` now, and create the `ResultSet` directly in PhoenixClient

lhofhansl replied:
Hmm... My $0.02. What leads to more code maintenance in the future is an interesting question.

Copying the code down allows base-jdbc to be changed with more ease it would perhaps also allow for using a few more of the "HBase'y" features of Phoenix (that might not map cleanly to JDBC);
on the other hand we now have copies of the code lying around, which now also have to be maintained, and base-jdbc already does a lot of the work that this connector would have to duplicate otherwise.

Personally I'd err on the side of less code-duplication. We should look at Phoenix connector as a JDBC type connector. That way things are simplified and improvements to base-jdbc would be visible everywhere.

kokosing replied:
In general, you are correct. My point was that if you are using API that is prepared for certain use case and your use case is very different then it may become difficult to maintain the API. Then changing API may benefit in one use case, but could be a problem in the other. However, it is always a trade-off and sometimes it is hard to tell what is better to reuse or to copy.

Here, I wanted to make sure we are not is such a place.

kokosing replied:
>  I removed usage of PhoenixRecordReader now, and create the ResultSet directly in PhoenixClient

Nice

vincentpoon commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,283 @@
+        }
+        catch (Exception e) {
+            try (Connection closingConnection = connection) {
+                // use try with resources to close properly
```
Just meant exactly that.  It's similar to the comment in JdbcRecordCursor#close()
Also, I think checkstyle requires something in the braces

kokosing replied:
I would update comment to `use try with resources to not to shade origin exception`

vincentpoon commented on .travis.yml:
```diff
@@ -19,7 +19,8 @@ env:
     - TEST_SPECIFIC_MODULES=presto-mongodb
-    - TEST_OTHER_MODULES=!presto-tests,!presto-raptor-legacy,!presto-accumulo,!presto-cassandra,!presto-hive,!presto-kudu,!presto-docs,!presto-server,!presto-server-rpm,!presto-main,!presto-mongodb
+    - TEST_SPECIFIC_MODULES=presto-phoenix
+    - TEST_OTHER_MODULES=!presto-tests,!presto-raptor-legacy,!presto-accumulo,!presto-cassandra,!presto-hive,!presto-kudu,!presto-phoenix,!presto-docs,!presto-server,!presto-server-rpm,!presto-main,!presto-mongodb
```
Not sure what the convention is here, as there are different ways to do it in YAML, and I'm not familiar with travis nuances.  So left this for now

vincentpoon commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,283 @@
+        try {
+            Field field = PhoenixRecordReader.class.getDeclaredField("resultSet");
+            field.setAccessible(true);
+            this.resultSetField = field;
```
I was trying to reuse some code in PhoenixRecordReader, but changed it now as I found I could just build the ResultSet in a helper method here.  Thanks

kokosing commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,314 @@
+    @Override
+    public Set<String> getSchemaNames(JdbcIdentity identity)
+    {
+        ImmutableSet.Builder<String> schemaNames = ImmutableSet.builder();
```
inline `schemaNames`

kokosing commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,314 @@
+            throws SQLException
+    {
+        return new SchemaTableName(
+                toPrestoSchemaName(Optional.ofNullable(resultSet.getString(TABLE_SCHEM))).orElse(null),
```
`SchemaTableName` constructor does not accept `null` parameters

vincentpoon commented:
Thanks @kokosing , updated to address your comments

combineads commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixMetadata.java:
```diff
@@ -0,0 +1,388 @@
+        return new ConnectorTableMetadata(handle.getSchemaTableName(), columnMetadata.build(), getTableProperties(session, handle));
+    }
+
+    public Map<String, Object> getTableProperties(ConnectorSession session, JdbcTableHandle handle)
```
rowkeys (with row_timestamp ) table properties isn't print from `show create table`.

vincentpoon replied:
@combineads Added support for showing rowkeys in the table properties

combineads commented:
This phoenix connector occurs the below error when using 4.14.0-cdh5.14.2 phoenix version.

```
java.lang.AbstractMethodError: io.prestosql.plugin.jdbc.JdbcRecordSetProvider.getRecordSet(Lio/prestosql/spi/connector/ConnectorTransactionHandle;Lio/prestosql/spi/connector/ConnectorSession;Lio/prestosql/spi/connector/ConnectorSplit;Ljava/util/List;)Lio/prestosql/spi/connector/RecordSet;
	at io.prestosql.split.RecordPageSourceProvider.createPageSource(RecordPageSourceProvider.java:42)
	at io.prestosql.split.PageSourceManager.createPageSource(PageSourceManager.java:56)
	at io.prestosql.operator.TableScanOperator.getOutput(TableScanOperator.java:239)
	at io.prestosql.operator.Driver.processInternal(Driver.java:379)
	at io.prestosql.operator.Driver.lambda$processFor$8(Driver.java:283)
	at io.prestosql.operator.Driver.tryWithLock(Driver.java:675)
	at io.prestosql.operator.Driver.processFor(Driver.java:276)
	at io.prestosql.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1075)
	at io.prestosql.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
	at io.prestosql.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:484)
	at io.prestosql.$gen.Presto_306____20190329_083613_1.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)```

Praveen2112 commented:
@combineads I guess you have to run it with the latest build of presto. I guess there is a recent changes added to the `RecordSet` API and old version of Presto (I assume this Presto-306 was running).

dain commented:
If this is ready for another review, can you resolve the conflict in the `.travis.yml` file, so travis will run?

combineads commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixConfig.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Config("phoenix.hbase-site-path")
+    public PhoenixConfig setConnectionProperties(String configPath)
```
For kerberos authentication, core-site.xml is also required. Could you provide support for putting multiple xml files in `phoenix.hbase-site-path` with commas? Right now I am manually merging and using configuration files.

lhofhansl commented:
I noticed now that when I add a table in Phoenix without a schema that SHOW SCHEMAS FROM <Phoenix-Catalog> throws an NPE.

Indeed BaseJdbcClient.getSchemaNames() expects that resultSet.getString("TABLE_SCHEM") always returns something non-null, which is not correct according to the JDBC documentation. (TABLE_SCHEM might be null).

```
java.lang.NullPointerException: undefined
	at io.prestosql.plugin.jdbc.BaseJdbcClient.getSchemaNames(BaseJdbcClient.java:129)
	at io.prestosql.plugin.phoenix.PhoenixClient.getSchemaNames(PhoenixClient.java:167)
	at io.prestosql.plugin.phoenix.PhoenixMetadata.listSchemaNames(PhoenixMetadata.java:102)
	at io.prestosql.metadata.MetadataManager.listSchemaNames(MetadataManager.java:313)
	at io.prestosql.connector.informationschema.InformationSchemaMetadata.calculatePrefixesWithSchemaName(InformationSchemaMetadata.java:277)
	at io.prestosql.connector.informationschema.InformationSchemaMetadata.getTableLayouts(InformationSchemaMetadata.java:241)
	at io.prestosql.metadata.MetadataManager.getLayout(MetadataManager.java:404)
	at io.prestosql.sql.planner.iterative.rule.PushPredicateIntoTableScan.pushFilterIntoTableScan(PushPredicateIntoTableScan.java:205)
	at io.prestosql.sql.planner.optimizations.AddExchanges$Rewriter.planTableScan(AddExchanges.java:540)
	at io.prestosql.sql.planner.optimizations.AddExchanges$Rewriter.visitTableScan(AddExchanges.java:507)
	at io.prestosql.sql.planner.optimizations.AddExchanges$Rewriter.visitTableScan(AddExchanges.java:133)
	at io.prestosql.sql.planner.plan.TableScanNode.accept(TableScanNode.java:131)
	at io.prestosql.sql.planner.optimizations.AddExchanges$Rewriter.planChild(AddExchanges.java:1166)
	at io.prestosql.sql.planner.optimizations.AddExchanges$Rewriter.visitSort(AddExchanges.java:413)
	at io.prestosql.sql.planner.optimizations.AddExchanges$Rewriter.visitSort(AddExchanges.java:133)
	at io.prestosql.sql.planner.plan.SortNode.accept(SortNode.java:74)
	at io.prestosql.sql.planner.optimizations.AddExchanges$Rewriter.planChild(AddExchanges.java:1166)
	at io.prestosql.sql.planner.optimizations.AddExchanges$Rewriter.visitOutput(AddExchanges.java:175)
	at io.prestosql.sql.planner.optimizations.AddExchanges$Rewriter.visitOutput(AddExchanges.java:133)
	at io.prestosql.sql.planner.plan.OutputNode.accept(OutputNode.java:82)
	at io.prestosql.sql.planner.optimizations.AddExchanges.optimize(AddExchanges.java:129)
	at io.prestosql.sql.planner.optimizations.StatsRecordingPlanOptimizer.optimize(StatsRecordingPlanOptimizer.java:51)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:170)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:159)
	at io.prestosql.execution.SqlQueryExecution.doAnalyzeQuery(SqlQueryExecution.java:423)
	at io.prestosql.execution.SqlQueryExecution.analyzeQuery(SqlQueryExecution.java:408)
	at io.prestosql.execution.SqlQueryExecution.startExecution(SqlQueryExecution.java:352)
	at io.prestosql.$gen.Presto_307_dirty____20190406_173206_1.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

@vincentpoon

kokosing commented on presto-docs/src/main/sphinx/connector/phoenix.rst:
```diff
@@ -0,0 +1,173 @@
+
+    connector.name=phoenix
+    phoenix.connection-url=jdbc:phoenix:host1,host2,host3:2181:/hbase
+    phoenix.hbase-site-path=/path/to/hbase-site.xml
```
for hive we are using `hive.config.resources` and this accepts a list of hadoop configuration files. Maybe here we could do similar `phoenix.config.resources`?

kokosing commented on presto-phoenix/pom.xml:
```diff
@@ -0,0 +1,175 @@
+        <dependency>
+            <groupId>org.apache.phoenix</groupId>
+            <artifactId>presto-phoenix-client-shaded</artifactId>
+            <version>4.14.1-HBase-1.4-PPS-0.3</version>
```
extract version as property and use it elsewhere

kokosing commented on presto-phoenix/pom.xml:
```diff
@@ -0,0 +1,175 @@
+        <dependency>
+            <groupId>org.apache.phoenix</groupId>
+            <artifactId>presto-phoenix-client-shaded</artifactId>
+            <version>4.14.1-HBase-1.4-PPS-0.3</version>
```
use version property (see comment above)

kokosing commented on presto-phoenix/pom.xml:
```diff
@@ -0,0 +1,175 @@
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>jcl-over-slf4j</artifactId>
+            <scope>runtime</scope>
```
please move somewhere near compile dependencies

kokosing commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/MetadataUtil.java:
```diff
@@ -0,0 +1,44 @@
+
+    public static Optional<String> toPrestoSchemaName(Optional<String> phoenixSchemaName)
+    {
+        return phoenixSchemaName.map(schemaName -> isNullOrEmpty(schemaName) ? DEFAULT_SCHEMA : schemaName);
```
I think Optional cannot have `null` inside

kokosing commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,376 @@
+        requireNonNull(config, "config is null");
+        hbaseConfig = new Configuration(false);
+        ConnectionInfo connectionInfo = PhoenixEmbeddedDriver.ConnectionInfo.create(config.getConnectionUrl());
+        connectionInfo.asProps().forEach(prop -> hbaseConfig.set(prop.getKey(), prop.getValue()));
```
I'd format this as:

```
connectionInfo.asProps()
    .forEach(prop -> hbaseConfig.set(prop.getKey(), prop.getValue()));
```

kokosing commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,376 @@
+public class PhoenixClient
+        extends BaseJdbcClient
+{
+    private final Configuration hbaseConfig;
```
s/hbaseConfig/configuration?

kokosing commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixMetadata.java:
```diff
@@ -0,0 +1,626 @@
+import static org.apache.phoenix.query.QueryConstants.SYSTEM_SCHEMA_NAME;
+import static org.apache.phoenix.util.PhoenixRuntime.getTable;
+
+public class PhoenixMetadata
```
Also it would be nice to able to somehow see what have changed comparing to base-jdbc

kokosing commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,155 @@
+            // We shouldn't really ever get here since getLocations() just returns a field value
+            // The location was already fetched as part of generateSplits()
+            log.warn("Failed to get host addresses for split: " + split, e);
+            return ImmutableList.of();
```
I don't think we should continue on `InterruptedException`

kokosing commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,155 @@
+            return ImmutableList.of(HostAddress.fromString(split.getLocations()[0]));
+        }
+        catch (IOException | InterruptedException e) {
+            // We shouldn't really ever get here since getLocations() just returns a field value
```
then please throw here an exception if we should never reach here.

kokosing commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixTableLayoutHandle.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.Optional;
+import java.util.Set;
+
+public class PhoenixTableLayoutHandle
```
TableLayout are deprecated please remove it.

vincentpoon replied:
After I asked in Slack #general, it was determined that we should continue using the deprecated APIs for now as there's no other way currently to get `desiredColumns`

kokosing commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/PhoenixQueryRunner.java:
```diff
@@ -0,0 +1,134 @@
+            throws Exception
+    {
+        DistributedQueryRunner queryRunner = DistributedQueryRunner.builder(createSession())
+                .setNodeCount(4)
```
this is default

kokosing commented on presto-phoenix/src/test/resources/hbase-site.xml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0"?>
```
This file looks unused. If I am wrong then it would be nice to configure that from test code via some phoenix config object.

kokosing commented:
I would also consider to squash all the commits and open a new PR. This one already has more than 200 of comments and github is getting slow.

vincentpoon commented:
Pushed another update that addresses the comments.  Also, this now uses the official phoenix-client, instead of the custom shaded jar we were producing for Presto.  It required a few changes to the pom.xml
Multiple resource files for configuration is now supported as well.

vincentpoon commented:
Closed this since GitHub was loading slowly with this many comments.  Continuing at #672
